cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go

```diff
@@ -108,12 +108,12 @@ func (ng *nodegroup) DeleteNodes(nodes []*corev1.Node) error {
 	}
 
 	// Step 2: if deleting len(nodes) would make the replica count
-	// <= 0, then the request to delete that many nodes is bogus
-	// and we fail fast.
+	// < node group minimum size, then the request to delete that many
+	// nodes is bogus and we fail fast.
 	replicas := ng.scalableResource.Replicas()
 
-	if replicas-int32(len(nodes)) < 0 {
-		return fmt.Errorf("unable to delete %d machines in %q, machine replicas are < 0 ", len(nodes), ng.Id())
+	if replicas-int32(len(nodes)) < int32(ng.MinSize()) {
```

Comment:
It's not clear to me why we're being passed too many nodes here in the first place.

Say we have minimum 1, current 4, it wants to delete 4.  Right now, that leaves nodes dangling with an annotation, but otherwise unaffected.  If we make this change, is the scale down ever going to happen?

I tried tracing back through the code, it's not really obvious how it's deciding which nodes to send in the list.  It looks like it's all of them for a given group.

Reply:
i don't think that this specific test will resolve the issue as reported, but while i was exploring the delete path it occurred to me that under some situations it might be possible for `replics - nodes` to equal 1 where a delete had already been scheduled. if that were to occur in a poorly timed manner, then i think 1 annotated node could _slip through_ as it were.

but, i felt like it was just safer to add the early check on a situation where we were requesting to go below the minimum size as that seemed like a good point to return.

Reply:
That doesn't really answer the quandary of "autoscaler wants to remove 2, only has room to remove 1."  Is it going to fail, and try to remove 1 on the next cycle, or is it always going to try to remove 2?

Reply:
indeed, that's a good point. i am digging through the autoscaler to better understand how this delete is happening. from what i was watching though, i'm thinking that instead of it getting a request for deleting 2 nodes, it is instead getting 2 requests to delete 1 node and the timing is making it appear that they happen at the same time. so, both are processing and in our final check one will fail as the other has already reduced the replica count.

at least, that's my working theory.